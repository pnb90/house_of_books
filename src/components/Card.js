import React from 'react'
import styled from 'styled-components'
import Dialog from './Dialog'

const Card = styled.div`
  width: 225px; 
  height: 300px;                
  margin: 12px;
  padding: 5px;
  position: relative;
`

const BookCover = styled.img`
  width: 100%;
  height: 100%;
  border-radius: 5px;
`

const Overlay = styled.div`
  position: absolute;
  width: 225px;
  height: 300px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  overflow: auto;
  align-items: center;
  opacity: 0;
  border-radius: 5px;
  padding: 20px;
  box-sizing: border-box;


  &:hover{
    transition: .25s ease;
    opacity: 1;
    background: rgba(51,105,30,0.8);
    color: white;
    font-size: 20px;
  }
`

function CardComponent(props) { 

  // Default values to cover edge cases where bookcovers or authors don't exist
  let bookCover = "https://images-na.ssl-images-amazon.com/images/I/91GQ%2BOWqgHL._SX425_.jpg"
  let authors = "N/A"
  let publisher = "N/A"
  let description = "N/A"
  
  if ( props.book.volumeInfo.imageLinks ) {
    bookCover = props.book.volumeInfo.imageLinks.thumbnail
  }

  if ( props.book.volumeInfo.publisher ) {
    publisher = props.book.volumeInfo.publisher 
  }
  
  if (( props.book.volumeInfo.description ) && ( props.book.volumeInfo.description != "" )) {
    description = props.book.volumeInfo.description
  }

  if ( props.book.volumeInfo.authors ) {
    authors = props.book.volumeInfo.authors
    
    if (props.book.volumeInfo.authors.length > 1) {
      authors = props.book.volumeInfo.authors.join(", ")
    }
  }



  return(
      <Card>
        <Overlay>
          <h3> { props.book.volumeInfo.title } </h3>
          By: { authors } <br />
          Publisher: { publisher } <br />
          <Dialog 
            book = { props.book.volumeInfo }
            extraInfo = { props.book }
            bookCover = { bookCover }
            authors = { authors }
            publisher = { publisher }
            description = { description }
          />
        </Overlay>
        <BookCover src={ bookCover } alt="book cover"/>
      </Card>
  )
}

export default CardComponent
