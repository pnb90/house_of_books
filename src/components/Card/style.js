import styled from 'styled-components'

export const Card = styled.div`
  display: flex;
  width: 16%; 
  height: 40%;                
  margin: 1.25em;
  position: relative;
`

export const BookCover = styled.img`
  width: 100%;
  height: 100%;
  border-radius: 5px;
`

export const Overlay = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  flex-direction: column;
  justify-content: center;
  overflow: auto;
  opacity: 0;
  border-radius: 5px;
  padding: 1em;
  box-sizing: border-box;

  &:hover{
    transition: .25s ease;
    opacity: 1;
    background: rgba(51,105,30,0.8);
    color: white;
    font-size: 20px;
  }
`

export const OverlayTitle = styled.h1`
  font-size: 1.66rem;
`

export const OverlayItems = styled.ul`
  font-size: 1.25rem;
  padding: 0;
`