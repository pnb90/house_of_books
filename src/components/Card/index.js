import React from 'react'
import Dialog from '../Dialog'
import PropTypes from 'prop-types'
import { Card, BookCover, Overlay, OverlayTitle, OverlayItems } from "./style"

/**
 * Renders a single card object that displays as a bookcover with an overlay component on top of the bookcover. 
 * 
 * @param {*} props - contains information about individual book from results of GoogleBooks API search. Includes book object, title, authors, bookcover, publisher, and description.
 */

function CardComponent(props) { 
  return(
    <Card>
        <Overlay>
            <OverlayTitle> { props.title } </OverlayTitle>
            <OverlayItems> By: { props.authors } </OverlayItems>
            <OverlayItems> Publisher: { props.publisher } </OverlayItems>
            <Dialog 
            title = { props.title }
            bookCover = { props.bookCover }
            authors = { props.authors }
            publisher = { props.publisher }
            description = { props.description }
            externalLink = { props.externalLink }
            />
        </Overlay>
        <BookCover src={ props.bookCover } alt = {props.title + "'s book cover"}/>
    </Card>
  )
}

Card.propTypes = {
  key: PropTypes.string,
  book: PropTypes.object,
  title: PropTypes.string,
  bookCover: PropTypes.string,
  publisher: PropTypes.string,
  description: PropTypes.string,
  authors: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array
    ])
}

export default CardComponent
