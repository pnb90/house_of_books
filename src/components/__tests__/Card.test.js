import React from 'react';
import { render } from '@testing-library/react'
import Card from '../Card';

describe('Card component', () => {  
   it("matches snapshot", () => {
    const { asFragment } = render(<Card />)
    expect(asFragment()).toMatchSnapshot()
   })

   it("renders visible props properly", () => {
      const { getByText } = render(<Card
         title = "Test Title"
         publisher = "Test Publisher"
         authors = "Test Author"
         />)
      
      expect(getByText("Test Title")).toBeTruthy()
      expect(getByText("Publisher: Test Publisher")).toBeTruthy()
      expect(getByText("By: Test Author")).toBeTruthy()
   })

   describe('BookCover rendering', () => {
      it("passes and renders src url accurately", () => {
         const { getByAltText } = render(<Card
            bookCover = "http://bookcover.com"
            title = "TestBook"
            />)
   
         let bookCoverImg = getByAltText("TestBook's book cover")
         
         expect(bookCoverImg).toBeTruthy()
         expect(bookCoverImg.src).toBeTruthy()
         expect(bookCoverImg.src).toBe("http://bookcover.com/")
      })
   
      it("passes and renders alt text accurately", () => {
         const { getByAltText } = render(<Card
            bookCover = "http://bookcover.com"
            title = "TestBook"
            />)
   
         let bookCoverImg = getByAltText("TestBook's book cover")
         
         expect(bookCoverImg).toBeTruthy()      
         expect(bookCoverImg.alt).toBeTruthy()
         expect(bookCoverImg.alt).toBe("TestBook's book cover")
      })
   })
})

