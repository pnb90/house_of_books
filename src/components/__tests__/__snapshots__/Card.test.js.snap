// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gzVnrw kTjCBY"
  >
    <div
      class="sc-dnqmqq kvtRn"
    >
      <h1
        class="sc-iwsKbI bquKru"
      >
          
      </h1>
      <ul
        class="sc-gZMcBi fUbFnV"
      >
         By:  
      </ul>
      <ul
        class="sc-gZMcBi fUbFnV"
      >
         Publisher:  
      </ul>
      <div>
        <button
          class="sc-ifAKCX iFmML"
        >
           More Info 
        </button>
      </div>
    </div>
    <img
      alt="undefined's book cover"
      class="sc-htoDjs iLrjRP"
    />
  </div>
</DocumentFragment>
`;
